{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a25aeecb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'HG02941', 'HG02976', 'HG03114', 'HG03135', 'HG03189', 'HG03271', 'HG03313', 'HG03367', 'HG02943', 'HG02977', 'HG02979', 'HG03115', 'HG03136', 'HG03190', 'HG03279', 'HG03280', 'HG03342', 'HG03369', 'HG02981', 'HG03117', 'HG03139', 'HG03291', 'HG03343', 'HG03370', 'HG02946', 'HG03118', 'HG03157', 'HG03195', 'HG03294', 'HG03351', 'HG03372', 'HG02947', 'HG03099', 'HG03120', 'HG03159', 'HG03196', 'HG03295', 'HG03352', 'HG03499', 'HG02952', 'HG03100', 'HG03121', 'HG03124', 'HG03160', 'HG03198', 'HG03297', 'HG03354', 'HG03511', 'HG02953', 'HG03103', 'HG03123', 'HG03126', 'HG03162', 'HG03199', 'HG03514', 'HG02968', 'HG03105', 'HG03127', 'HG03163', 'HG03202', 'HG03300', 'HG03515', 'HG02970', 'HG03108', 'HG03129', 'HG03166', 'HG03265', 'HG03301', 'HG03517', 'HG02922', 'HG02971', 'HG03109', 'HG03130', 'HG03168', 'HG03169', 'HG03267', 'HG03303', 'HG03518', 'HG02923', 'HG02973', 'HG03111', 'HG03172', 'HG03268', 'HG03304', 'HG03363', 'HG02938', 'HG02974', 'HG03112', 'HG03133', 'HG02944', 'HG03132', 'HG03193', 'HG03298', 'HG03520', 'HG03175', 'HG03270', 'HG03311', 'HG03366', 'HG03521']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 62%|██████████████████████████████████████▍                       | 1048598/1690514 [2:40:19<1:24:14, 127.00it/s]"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import argparse\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "# TODO are we interested in:\n",
    "# 'non_coding_transcript_variant', 'frameshift_variant', or 'PolyPhenCat' 'Unknown' category?\n",
    "cats = {\n",
    "    \"LoF\": np.array([\"LoF_HC\", \"LoF_LC\"]),\n",
    "    \"Consequence\": np.array(['3_prime_UTR_variant',\n",
    "                     '5_prime_UTR_variant',\n",
    "                     'TF_binding_site_variant',\n",
    "                     'downstream_gene_variant',\n",
    "                     'intergenic_variant',\n",
    "                     'intron_variant',\n",
    "                     'missense_variant',\n",
    "                     'regulatory_region_variant',\n",
    "                     'splice_acceptor_variant',\n",
    "                     'splice_donor_variant',\n",
    "                     'splice_region_variant',\n",
    "                     'stop_gained',\n",
    "                     'synonymous_variant',\n",
    "                     'upstream_gene_variant',\n",
    "                     'frameshift_variant',\n",
    "                    ]),\n",
    "}\n",
    "\n",
    "\n",
    "def to_cat(var, cat):\n",
    "    if len(var) and cat == 'LoF': var = f'LoF_{var}'\n",
    "    return pd.DataFrame(np.isin(cats[cat],var).astype(int).reshape(1,-1), columns=cats[cat])\n",
    "\n",
    "def get_header():\n",
    "    header = [\"SubjectID\", \"GeneName\", \"Chromosome\", \"Position\", \"Ref\", \"Alt\", \"AF\"]\n",
    "    header.extend(cats[\"Consequence\"])\n",
    "    header.extend(cats[\"LoF\"])\n",
    "    return \"\\t\".join(header) + \"\\n\"\n",
    "\n",
    "def filelen(path):\n",
    "    fname = path.split(\"/\")[-1]\n",
    "    os.system(f\"wc -l {path} > /tmp/wc_{fname}\")\n",
    "    n = int(open(f\"/tmp/wc_{fname}\", \"r\").readlines()[0].split()[0])\n",
    "    return n    \n",
    "\n",
    "def get_vep(l):\n",
    "    if 'CSQ=' not in l[7]: return None\n",
    "\n",
    "    # TODO read this from the file, right now I am stripping that part of the header.\n",
    "    vep_fields = 'Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|LoF|LoF_filter|LoF_flags|LoF_info'.split(\"|\")\n",
    "    lidx = np.where(np.array(vep_fields) == \"LoF\")[0][0]\n",
    "    vidx = np.where(np.array(vep_fields) == \"Consequence\")[0][0]\n",
    "    gidx = np.where(np.array(vep_fields) == \"Gene\")[0][0]\n",
    "    \n",
    "    vep_infos = l[7].split('CSQ=')[1].split(',')\n",
    "    vep_infos = [v.split('|') for v in vep_infos]\n",
    "\n",
    "    vep_df = pd.DataFrame(vep_infos, columns=vep_fields)\n",
    "\n",
    "    # convert to categorical\n",
    "    vep_cat_df = pd.DataFrame()\n",
    "    for v in vep_df.index:\n",
    "        var = to_cat(vep_df.loc[v][\"Consequence\"], \"Consequence\")\n",
    "        lof = to_cat(vep_df.loc[v][\"LoF\"], \"LoF\")\n",
    "        cats = pd.concat([var, lof], axis=1)\n",
    "        vep_cat_df = pd.concat([vep_cat_df, cats], axis=0)\n",
    "    vep_cat_df.index = vep_df.loc[:,\"Gene\"]\n",
    "\n",
    "    # aggregate multiple annotations over same gene using max\n",
    "    for g in vep_cat_df.index.unique():\n",
    "        if len(vep_cat_df.loc[[g]]) > 1:\n",
    "            df2 = pd.DataFrame(vep_cat_df.loc[[g]].max(0)).T\n",
    "            df2.index = [g]\n",
    "            vep_cat_df = pd.concat([vep_cat_df.drop(g), df2])\n",
    "\n",
    "    return vep_cat_df\n",
    "\n",
    "\n",
    "def process_line(l, cols):\n",
    "    \n",
    "    l = l.strip(\"\\n\").split()\n",
    "    rare_idx = np.where(np.isin(l, [\"0/1\", \"1/0\", \"1/1\", \"0|1\", \"1|0\", \"1|1\"]))[0]\n",
    "    rare_ids = [cols[i] for i in rare_idx]\n",
    "    af = l[7].split(\"AFR_AF=\")[1].split(\";\")[0]\n",
    "    chidx = np.where(np.array(cols) == \"CHROM\")[0][0]\n",
    "    pidx = np.where(np.array(cols) == \"POS\")[0][0]\n",
    "    refidx = np.where(np.array(cols) == \"REF\")[0][0]\n",
    "    altidx = np.where(np.array(cols) == \"ALT\")[0][0]\n",
    "    ch,pos,ref,alt = l[chidx], l[pidx], l[refidx], l[altidx]\n",
    "\n",
    "    vep_cat_df = get_vep(l)\n",
    "    if vep_cat_df is None: return None\n",
    "\n",
    "    # Create annotation line for each id, gene pair\n",
    "    out_lines = []\n",
    "    for rid in set(rare_ids):\n",
    "        for gene in vep_cat_df.index:\n",
    "            out_line = [rid, gene, ch, pos, ref, alt, af]\n",
    "            out_line.extend(vep_cat_df.loc[gene].array.astype(str).tolist())\n",
    "            out_lines.append(\"\\t\".join(out_line) + \"\\n\")\n",
    "\n",
    "    genes = vep_cat_df.index.array\n",
    "    return \"\".join(out_lines), genes, rare_ids\n",
    "\n",
    "    \n",
    "def vcf_to_tsv(in_file, out_dset, out_gene_outliers):\n",
    "    \n",
    "    vcf = open(in_file, 'r')\n",
    "    cols = vcf.readline().strip(\"#\").split()\n",
    "    print(cols)\n",
    "    n = filelen(in_file)    \n",
    "  \n",
    "    out = open(out_dset, 'w')\n",
    "    gout = open(out_gene_outliers, 'w')\n",
    "               \n",
    "    header = get_header()\n",
    "    out.write(header)\n",
    "\n",
    "    for l in tqdm(vcf, total=n):\n",
    "        out_line, genes, rid = process_line(l, cols)\n",
    "        if out_line is not None and out_line != \"\":\n",
    "            out.write(out_line)\n",
    "            for gene in genes:\n",
    "                gout.write(\"\\t\".join([gene] + rid) + \"\\n\")\n",
    "\n",
    "        \n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    \n",
    "#     argParser = argparse.ArgumentParser()\n",
    "#     argParser.add_argument(\"--pop\", default=\"AFR\", type=str)\n",
    "#     argParser.add_argument(\"--vep_dir\", default='/oak/stanford/groups/smontgom/erobb/data/vep', type=str)\n",
    "#     argParser.add_argument(\"--out_dir\", default='/oak/stanford/groups/smontgom/erobb/data/watershed', type=str)\n",
    "#     args = argParser.parse_args()\n",
    "\n",
    "#     tsv_out = f'{args.out_dir}/AF.all.{args.pop}.hg38a.ID.ba.VEP.rare.ws.tsv'\n",
    "#     gene_outliers = f'f\"{args.out_dir}/gene_outliers_{args.pop}.tsv'\n",
    "#     vcf_in = f\"{args.vep_dir}/AF.all.{args.pop}.hg38a.ID.ba.VEP.rare.vcf\"\n",
    "    \n",
    "    pop = 'ESN'\n",
    "    vep_dir = '/oak/stanford/groups/smontgom/erobb/data/vep'\n",
    "    out_dir = \"/tmp\" \n",
    "    tsv_out = f'{out_dir}/AF.all.{pop}.hg38a.ID.ba.VEP.rare.ws.tsv'\n",
    "    gene_outliers = f'{out_dir}/gene_outliers_{pop}.tsv'\n",
    "    vcf_in = f\"{vep_dir}/AF.all.{pop}.hg38a.ID.ba.VEP.rare.vcf\"\n",
    "    \n",
    "    vcf_to_tsv(vcf_in, tsv_out, gene_outliers)\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cfe2e51",
   "metadata": {},
   "source": [
    "### filter out IDs with too much variance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8604f5",
   "metadata": {},
   "source": [
    "### cadd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e808433",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cats = {\n",
    "    # TODO these are strange, why does PolyPhen include 'benign' and 'unknown' as categories but SIFT doesn't\n",
    "    \"SIFTcat\": np.array([\"deleterious\", \"tolerated\"]),\n",
    "    \"PolyPhenCat\": np.array([\"benign\", \"possibly_damaging\", \"unknown\"]),\n",
    "    \"LoF\": np.array([\"LoF_HC\", \"LoF_LC\"]),\n",
    "    \"Consequence\": np.array(['3_prime_UTR_variant',\n",
    "                     '5_prime_UTR_variant',\n",
    "                     'TF_binding_site_variant',\n",
    "                     'downstream_gene_variant',\n",
    "                     'intergenic_variant',\n",
    "                     'intron_variant',\n",
    "                     'missense_variant',\n",
    "#                      'non_coding_transcript_variant',\n",
    "#                      'non_coding_transcript_exon_variant',\n",
    "                     'regulatory_region_variant',\n",
    "                     'splice_acceptor_variant',\n",
    "                     'splice_donor_variant',\n",
    "                     'splice_region_variant',\n",
    "                     'stop_gained',\n",
    "                     'synonymous_variant',\n",
    "                     'upstream_gene_variant',\n",
    "                     'frameshift_variant',\n",
    "                    ]),\n",
    "}\n",
    "\n",
    "\n",
    "def query_tabix(ch, s, e, opt=\"\"):\n",
    "    os.system(f\"tabix -h {CADD_FILE} {ch}:{s}-{e} > /tmp/{ch}:{s}-{e}.tsv\")\n",
    "    return pd.read_table(f\"/tmp/{ch}:{s}-{e}.tsv\", header=1)\n",
    "\n",
    "\n",
    "def get_cadd(ch,pos,ref,alt):\n",
    "    # Only supports SNVs at this time\n",
    "    if not (len(ref) == 1 and len(alt)== 1):\n",
    "        return [\"\"] * len(cadd_anno)\n",
    "\n",
    "    # Query tabix\n",
    "    cadd_table = query_tabix(ch, pos, int(pos))\n",
    "    cadd_table = cadd_table[cadd_table[\"Alt\"] == alt][cadd_anno]\n",
    "    if not len(cadd_table):\n",
    "        return [\"\"] * len(cadd_anno)\n",
    "\n",
    "    # TODO pick the most impactful (min or max) for each variant\n",
    "    cadd = cadd_table.iloc[0].array.astype(str).tolist()\n",
    "\n",
    "    # Replace \"nan\" with \"\"\n",
    "    svi = np.where(cadd_table.columns == \"SIFTval\")[0][0]\n",
    "    if cadd[svi] == \"nan\": cadd[svi] = \"\"\n",
    "    pvi = np.where(cadd_table.columns == \"PolyPhenVal\")[0][0]\n",
    "    if cadd[pvi] == \"nan\": cadd[pvi] = \"\"\n",
    "\n",
    "    # Replace SIFT with categoriacal\n",
    "    sci = np.where(cadd_table.columns == \"SIFTcat\")[0][0]\n",
    "    sc = to_cat(cadd[sci], \"SIFTcat\").values.reshape(-1).astype(str).tolist()\n",
    "    cadd[sci] = sc[0]\n",
    "    for e in range(len(sc[1:])): cadd.insert(sci+e,sc[e])\n",
    "\n",
    "    # Replace PolyPhen with categorical    \n",
    "    pci = np.where(cadd_table.columns == \"PolyPhenCat\")[0][0] + len(sc[1:])\n",
    "    pc = to_cat(cadd[pci], \"PolyPhenCat\").values.reshape(-1).astype(str).tolist()\n",
    "    cadd[pci]=pc[0]\n",
    "    for e in range(len(pc[1:])): cadd.insert(pci+e,pc[e])\n",
    "    return cadd\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    argParser = argparse.ArgumentParser()\n",
    "    argParser.add_argument(\"--pop\", default=\"AFR\", type=str)\n",
    "    argParser.add_argument(\"--data_dir\", default='/oak/stanford/groups/smontgom/erobb/data/watershed', type=str)\n",
    "    argParser.add_argument(\"--cadd\", default=\"/oak/stanford/groups/smontgom/erobb/data/watershed/whole_genome_SNVs_inclAnno.tsv.gz\", type=str)\n",
    "    args = argParser.parse_args()\n",
    "    \n",
    "    cadd_anno = [\"GC\", \"CpG\", \"SIFTcat\", \"SIFTval\",\"PolyPhenCat\",\n",
    "             \"PolyPhenVal\", \"bStatistic\", \"priPhCons\",\"mamPhCons\",\"verPhCons\",\n",
    "             \"priPhyloP\",\"mamPhyloP\",\"verPhyloP\",\"GerpN\",\"GerpS\",\"PHRED\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf9b4b9",
   "metadata": {},
   "source": [
    "### eOutlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ec9cf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    #eout_file = f\"{data_dir}/eOutlier_scores_{pop}_t3.txt\"\n",
    "    eout_file = f\"/oak/stanford/groups/smontgom/erobb/data/watershed/{pop}_exprResiduals.tsv\"\n",
    "    eout_keep_file = f\"{data_dir}/ids_outlier_filtered_{pop}_t3f75.txt\"\n",
    "\n",
    "    # Expression outlier scores from residuals file\n",
    "    eout_file = f\"/oak/stanford/groups/smontgom/erobb/data/watershed/{pop}_exprResiduals.tsv\"\n",
    "    eout_keep_file = f\"{data_dir}/ids_outlier_filtered_{pop}_t3f75.txt\"\n",
    "    eout_df = pd.read_table(eout_file, sep=\"\\t\", index_col=0).T\n",
    "    # drop transcript information\n",
    "    gene_names = [c.split(\".\")[0] for c in eout_df.columns]\n",
    "    assert(len(np.unique(gene_names)) == len(eout_df.columns))\n",
    "    eout_df.columns = gene_names\n",
    "    inds_keep = pd.read_table(eout_keep_file, sep=\" \", index_col=0, header=None).T\n",
    "    eout_df = eout_df.loc[eout_df.index.intersection(inds_keep.columns)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28493e3f",
   "metadata": {},
   "source": [
    "### gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b4cf9a66",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                  | 0/650090 [00:00<?, ?it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/tmp/ipykernel_51291/1069628940.py\u001b[0m(42)\u001b[0;36madd_gencode\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m     40 \u001b[0;31m        \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     41 \u001b[0;31m    \u001b[0;32mimport\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m---> 42 \u001b[0;31m    \u001b[0mann\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcatenate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mann\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     43 \u001b[0;31m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m     44 \u001b[0;31m    \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"distTSS\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mann\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> ann\n",
      "[[nan, nan]]\n",
      "ipdb> df.iloc[:1]\n",
      "                    SubjectID         GeneName  Chromosome  Position    AF  \\\n",
      "Chromosome Position                                                          \n",
      "1          597782     HG03265  ENSG00000230021           1    597782  0.01   \n",
      "\n",
      "                     3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
      "Chromosome Position                                             \n",
      "1          597782                      0                    0   \n",
      "\n",
      "                     TF_binding_site_variant  downstream_gene_variant  \\\n",
      "Chromosome Position                                                     \n",
      "1          597782                          0                        1   \n",
      "\n",
      "                     intergenic_variant  ...  priPhCons  mamPhCons  verPhCons  \\\n",
      "Chromosome Position                      ...                                    \n",
      "1          597782                     0  ...      0.046      0.046      0.297   \n",
      "\n",
      "                     priPhyloP  mamPhyloP  verPhyloP  GerpN  GerpS  PHRED  \\\n",
      "Chromosome Position                                                         \n",
      "1          597782          0.0        0.0        0.0    0.0    0.0  5.241   \n",
      "\n",
      "                     eOutlier  \n",
      "Chromosome Position            \n",
      "1          597782         0.0  \n",
      "\n",
      "[1 rows x 42 columns]\n",
      "ipdb>  ann = pd.concatenate(ann)\n",
      "*** AttributeError: module 'pandas' has no attribute 'concatenate'\n",
      "ipdb> pd.concat(ann)\n",
      "*** TypeError: cannot concatenate object of type '<class 'list'>'; only Series and DataFrame objs are valid\n",
      "ipdb> pd.concat([np.array(ann[0])])\n",
      "*** TypeError: cannot concatenate object of type '<class 'numpy.ndarray'>'; only Series and DataFrame objs are valid\n",
      "ipdb> np.array(ann)\n",
      "array([[nan, nan]])\n",
      "ipdb>     ann = np.array(ann)\n",
      "ipdb> df.iloc[:1][\"distTSS\"] = ann[:,0]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<stdin>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ipdb> df\n",
      "                    SubjectID         GeneName  Chromosome  Position    AF  \\\n",
      "Chromosome Position                                                          \n",
      "1          597782     HG03265  ENSG00000230021           1    597782  0.01   \n",
      "           597782     HG03267  ENSG00000230021           1    597782  0.01   \n",
      "           597782     HG02976  ENSG00000230021           1    597782  0.01   \n",
      "           630096     HG03301  ENSG00000237973           1    630096  0.01   \n",
      "           630096     HG03301  ENSG00000198744           1    630096  0.01   \n",
      "...                       ...              ...         ...       ...   ...   \n",
      "5          60891442   HG02973  ENSG00000049167           5  60891442  0.01   \n",
      "           60893612   HG03372  ENSG00000049167           5  60893612  0.01   \n",
      "           60893612   HG03297  ENSG00000049167           5  60893612  0.01   \n",
      "           60893612   HG03117  ENSG00000049167           5  60893612  0.01   \n",
      "           60893612   HG03366  ENSG00000049167           5  60893612  0.01   \n",
      "\n",
      "                     3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
      "Chromosome Position                                             \n",
      "1          597782                      0                    0   \n",
      "           597782                      0                    0   \n",
      "           597782                      0                    0   \n",
      "           630096                      0                    0   \n",
      "           630096                      0                    0   \n",
      "...                                  ...                  ...   \n",
      "5          60891442                    0                    0   \n",
      "           60893612                    0                    0   \n",
      "           60893612                    0                    0   \n",
      "           60893612                    0                    0   \n",
      "           60893612                    0                    0   \n",
      "\n",
      "                     TF_binding_site_variant  downstream_gene_variant  \\\n",
      "Chromosome Position                                                     \n",
      "1          597782                          0                        1   \n",
      "           597782                          0                        1   \n",
      "           597782                          0                        1   \n",
      "           630096                          0                        0   \n",
      "           630096                          0                        0   \n",
      "...                                      ...                      ...   \n",
      "5          60891442                        0                        1   \n",
      "           60893612                        0                        1   \n",
      "           60893612                        0                        1   \n",
      "           60893612                        0                        1   \n",
      "           60893612                        0                        1   \n",
      "\n",
      "                     intergenic_variant  ...  priPhCons  mamPhCons  verPhCons  \\\n",
      "Chromosome Position                      ...                                    \n",
      "1          597782                     0  ...      0.046      0.046      0.297   \n",
      "           597782                     0  ...      0.046      0.046      0.297   \n",
      "           597782                     0  ...      0.046      0.046      0.297   \n",
      "           630096                     0  ...      0.007      0.007      0.007   \n",
      "           630096                     0  ...      0.007      0.007      0.007   \n",
      "...                                 ...  ...        ...        ...        ...   \n",
      "5          60891442                   0  ...      0.682      0.001      0.001   \n",
      "           60893612                   0  ...      0.162      0.471      0.477   \n",
      "           60893612                   0  ...      0.162      0.471      0.477   \n",
      "           60893612                   0  ...      0.162      0.471      0.477   \n",
      "           60893612                   0  ...      0.162      0.471      0.477   \n",
      "\n",
      "                     priPhyloP  mamPhyloP  verPhyloP  GerpN  GerpS  PHRED  \\\n",
      "Chromosome Position                                                         \n",
      "1          597782        0.000      0.000      0.000  0.000  0.000  5.241   \n",
      "           597782        0.000      0.000      0.000  0.000  0.000  5.241   \n",
      "           597782        0.000      0.000      0.000  0.000  0.000  5.241   \n",
      "           630096        0.000      0.000      0.000  0.000  0.000  3.492   \n",
      "           630096        0.000      0.000      0.000  0.000  0.000  3.492   \n",
      "...                        ...        ...        ...    ...    ...    ...   \n",
      "5          60891442      0.321      0.504      0.487  4.210  1.800  9.267   \n",
      "           60893612      0.404      0.698      0.675  0.642  0.642  7.065   \n",
      "           60893612      0.404      0.698      0.675  0.642  0.642  7.065   \n",
      "           60893612      0.404      0.698      0.675  0.642  0.642  7.065   \n",
      "           60893612      0.404      0.698      0.675  0.642  0.642  7.065   \n",
      "\n",
      "                     eOutlier  \n",
      "Chromosome Position            \n",
      "1          597782         0.0  \n",
      "           597782         0.0  \n",
      "           597782         0.0  \n",
      "           630096         0.0  \n",
      "           630096         0.0  \n",
      "...                       ...  \n",
      "5          60891442       0.0  \n",
      "           60893612       0.0  \n",
      "           60893612       0.0  \n",
      "           60893612       0.0  \n",
      "           60893612       0.0  \n",
      "\n",
      "[650090 rows x 42 columns]\n",
      "ipdb> dfp = df.iloc[:1]\n",
      "ipdb> dfp\n",
      "                    SubjectID         GeneName  Chromosome  Position    AF  \\\n",
      "Chromosome Position                                                          \n",
      "1          597782     HG03265  ENSG00000230021           1    597782  0.01   \n",
      "\n",
      "                     3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
      "Chromosome Position                                             \n",
      "1          597782                      0                    0   \n",
      "\n",
      "                     TF_binding_site_variant  downstream_gene_variant  \\\n",
      "Chromosome Position                                                     \n",
      "1          597782                          0                        1   \n",
      "\n",
      "                     intergenic_variant  ...  priPhCons  mamPhCons  verPhCons  \\\n",
      "Chromosome Position                      ...                                    \n",
      "1          597782                     0  ...      0.046      0.046      0.297   \n",
      "\n",
      "                     priPhyloP  mamPhyloP  verPhyloP  GerpN  GerpS  PHRED  \\\n",
      "Chromosome Position                                                         \n",
      "1          597782          0.0        0.0        0.0    0.0    0.0  5.241   \n",
      "\n",
      "                     eOutlier  \n",
      "Chromosome Position            \n",
      "1          597782         0.0  \n",
      "\n",
      "[1 rows x 42 columns]\n",
      "ipdb> dfp.insert(4,ann)\n",
      "*** TypeError: DataFrame.insert() missing 1 required positional argument: 'value'\n",
      "ipdb> df.insert(4, [\"distTSS\", \"distTES\"], ann)\n",
      "*** TypeError: unhashable type: 'list'\n",
      "ipdb>  df.insert(4, \"TSS\", ann)\n",
      "*** ValueError: Length of values (1) does not match length of index (650090)\n",
      "ipdb> dfp.insert(4, \"TSS\", ann)\n",
      "*** ValueError: Expected a 1D array, got an array with shape (1, 2)\n",
      "ipdb> dfp.insert(4, \"TSS\", ann[:,0])\n",
      "ipdb> dfp\n",
      "                    SubjectID         GeneName  Chromosome  Position  TSS  \\\n",
      "Chromosome Position                                                         \n",
      "1          597782     HG03265  ENSG00000230021           1    597782  NaN   \n",
      "\n",
      "                       AF  3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
      "Chromosome Position                                                   \n",
      "1          597782    0.01                    0                    0   \n",
      "\n",
      "                     TF_binding_site_variant  downstream_gene_variant  ...  \\\n",
      "Chromosome Position                                                    ...   \n",
      "1          597782                          0                        1  ...   \n",
      "\n",
      "                     priPhCons  mamPhCons  verPhCons  priPhyloP  mamPhyloP  \\\n",
      "Chromosome Position                                                          \n",
      "1          597782        0.046      0.046      0.297        0.0        0.0   \n",
      "\n",
      "                     verPhyloP  GerpN  GerpS  PHRED  eOutlier  \n",
      "Chromosome Position                                            \n",
      "1          597782          0.0    0.0    0.0  5.241       0.0  \n",
      "\n",
      "[1 rows x 43 columns]\n",
      "ipdb> dfp.insert(5, \"distTES\", ann[:,1])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ipdb> dfp\n",
      "                    SubjectID         GeneName  Chromosome  Position  TSS  \\\n",
      "Chromosome Position                                                         \n",
      "1          597782     HG03265  ENSG00000230021           1    597782  NaN   \n",
      "\n",
      "                     distTES    AF  3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
      "Chromosome Position                                                            \n",
      "1          597782        NaN  0.01                    0                    0   \n",
      "\n",
      "                     TF_binding_site_variant  ...  priPhCons  mamPhCons  \\\n",
      "Chromosome Position                           ...                         \n",
      "1          597782                          0  ...      0.046      0.046   \n",
      "\n",
      "                     verPhCons  priPhyloP  mamPhyloP  verPhyloP  GerpN  GerpS  \\\n",
      "Chromosome Position                                                             \n",
      "1          597782        0.297        0.0        0.0        0.0    0.0    0.0   \n",
      "\n",
      "                     PHRED  eOutlier  \n",
      "Chromosome Position                   \n",
      "1          597782    5.241       0.0  \n",
      "\n",
      "[1 rows x 44 columns]\n",
      "--KeyboardInterrupt--\n",
      "\n",
      "KeyboardInterrupt: Interrupted by user\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "module 'pandas' has no attribute 'concatenate'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 72\u001b[0m\n\u001b[1;32m     68\u001b[0m var_df \u001b[38;5;241m=\u001b[39m var_df\u001b[38;5;241m.\u001b[39msort_values([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChromosome\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPosition\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m     69\u001b[0m var_df \u001b[38;5;241m=\u001b[39m var_df\u001b[38;5;241m.\u001b[39mset_index([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mChromosome\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPosition\u001b[39m\u001b[38;5;124m\"\u001b[39m], drop\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m---> 72\u001b[0m gdf \u001b[38;5;241m=\u001b[39m \u001b[43madd_gencode\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvar_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgencode\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[37], line 42\u001b[0m, in \u001b[0;36madd_gencode\u001b[0;34m(df, gencode)\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpdb\u001b[39;00m; pdb\u001b[38;5;241m.\u001b[39mset_trace()\n\u001b[0;32m---> 42\u001b[0m ann \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconcatenate\u001b[49m(ann)\n\u001b[1;32m     44\u001b[0m df\u001b[38;5;241m.\u001b[39minsert(\u001b[38;5;241m4\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdistTSS\u001b[39m\u001b[38;5;124m\"\u001b[39m, ann[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues)\n\u001b[1;32m     45\u001b[0m df\u001b[38;5;241m.\u001b[39minsert(\u001b[38;5;241m5\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdistTES\u001b[39m\u001b[38;5;124m\"\u001b[39m, ann[\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues)\n",
      "File \u001b[0;32m~/miniconda3/envs/ws/lib/python3.11/site-packages/pandas/__init__.py:264\u001b[0m, in \u001b[0;36m__getattr__\u001b[0;34m(name)\u001b[0m\n\u001b[1;32m    260\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marrays\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01msparse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m SparseArray \u001b[38;5;28;01mas\u001b[39;00m _SparseArray\n\u001b[1;32m    262\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _SparseArray\n\u001b[0;32m--> 264\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodule \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpandas\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m has no attribute \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'pandas' has no attribute 'concatenate'"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import argparse\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "def binsearch(df, pos, c1, c2):\n",
    "    \"\"\"\n",
    "    Assumes dataframe is sorted by c1,c2.\n",
    "    \"\"\"\n",
    "    n = len(df)\n",
    "    if n == 0: return None\n",
    "    if pos > df[c2].iloc[n//2]: # pos is to the right of \"end\"\n",
    "        return binsearch(df.iloc[n//2+1:], pos, c1, c2)\n",
    "    elif pos < df[c1].iloc[n//2]: # pos is to the left of \"start\"\n",
    "        return binsearch(df.iloc[:n//2], pos, c1, c2)\n",
    "    elif df[c1].iloc[n//2] <= pos and df[c2].iloc[n//2] >= pos:\n",
    "        return df.iloc[n//2]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def add_gencode(df, gencode):\n",
    "    gc = pd.read_table(gencode, header=None)\n",
    "    gc = gc.sort_values([0,3,4])\n",
    "    gc = gc.set_index([0],drop=False)\n",
    "    \n",
    "    def _query(row):\n",
    "        ch, pos = row.index.unique()[0]\n",
    "        gc_ch_pos = binsearch(gc.loc[f\"chr{ch}\"], pos, 3, 4)\n",
    "        if gc_ch_pos is None: return [np.nan, np.nan]\n",
    "        TSS = pos - gc_ch_pos[3] #gc.loc[w,3].values[0]\n",
    "        TES = gc_ch_pos[4] - pos #gc.loc[w,4].values[0] - pos\n",
    "        return [TSS, TES]\n",
    "\n",
    "    ann = []\n",
    "    for idx in tqdm(df.index):\n",
    "        ann.append(_query(df.loc[idx]))\n",
    "    ann = np.array(ann)\n",
    "\n",
    "    df.insert(4, \"distTSS\", ann[:,0])\n",
    "    df.insert(5, \"distTES\", ann[:,1])\n",
    "    return df\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    argParser = argparse.ArgumentParser()\n",
    "    # argParser.add_argument(\"--pop\", default=\"ESN\", type=str)\n",
    "    # argParser.add_argument(\"--data_dir\", default='/oak/stanford/groups/smontgom/erobb/data/watershed', type=str)\n",
    "    # argParser.add_argument(\"--gencode\", default=\"/oak/stanford/groups/smontgom/erobb/data/watershed/gencode.v43.chr_patch_hapl_scaff.annotation.exons.protein_lincRNA.gtf\", type=str)\n",
    "    #args = argParser.parse_args()\n",
    "    # pop = args.pop\n",
    "    \n",
    "    pop = \"ESN\"\n",
    "    gencode = \"/oak/stanford/groups/smontgom/erobb/data/watershed/gencode.v43.chr_patch_hapl_scaff.annotation.exons.protein_lincRNA.gtf\"\n",
    "    data_dir = '/oak/stanford/groups/smontgom/erobb/data/watershed'\n",
    "    tsv_in = f'AF.all.{pop}.hg38a.ID.ba.VEP.rare.ws.tsv'\n",
    "    tsv_out =  f'AF.all.{pop}.hg38a.ID.ba.VEP.gencode.rare.ws.tsv'\n",
    "    tsv_file = f'{data_dir}/{tsv_in}'\n",
    "    tsv_file_out = f'{data_dir}/{tsv_out}'\n",
    "\n",
    "\n",
    "    var_df = pd.read_table(tsv_file)\n",
    "    var_df = var_df.sort_values(['Chromosome', 'Position'])\n",
    "    var_df = var_df.set_index([\"Chromosome\", \"Position\"], drop=False)\n",
    "\n",
    "    \n",
    "    gdf = add_gencode(var_df, gencode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1f0c71f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "597782\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284733.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>923923</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>923923</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925150</td>\n",
       "      <td>925189</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925731</td>\n",
       "      <td>925800</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925922</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925942</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>925942</td>\n",
       "      <td>926013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187634.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0       1     2       3       4  5  6  7  \\\n",
       "0                                                   \n",
       "chr1  chr1  HAVANA  exon   65520   65573  .  +  .   \n",
       "chr1  chr1  HAVANA  exon   69037   71585  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  450740  451678  .  -  .   \n",
       "chr1  chr1  HAVANA  exon  685716  686654  .  -  .   \n",
       "chr1  chr1  HAVANA  exon  923923  924948  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  923923  924948  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925150  925189  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925731  925800  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925922  926013  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925922  926013  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925922  926013  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925922  926013  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925942  926013  .  +  .   \n",
       "chr1  chr1  HAVANA  exon  925942  926013  .  +  .   \n",
       "\n",
       "                                                      8  \n",
       "0                                                        \n",
       "chr1  gene_id \"ENSG00000186092.7\"; transcript_id \"EN...  \n",
       "chr1  gene_id \"ENSG00000186092.7\"; transcript_id \"EN...  \n",
       "chr1  gene_id \"ENSG00000284733.2\"; transcript_id \"EN...  \n",
       "chr1  gene_id \"ENSG00000284662.2\"; transcript_id \"EN...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  \n",
       "chr1  gene_id \"ENSG00000187634.13\"; transcript_id \"E...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(597782)\n",
    "#gc = pd.read_table(gencode, header=None)\n",
    "gc = gc.sort_values([3,4])\n",
    "gc.loc[\"chr1\"].iloc[1:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "dbe85f28",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos=597782\n",
    "ch=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "749bd1ac",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 0 is out of bounds for axis 0 with size 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[81], line 5\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# gc = pd.read_table(gencode, header=None)\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# gc = gc.set_index([0,3,4],drop=False)\u001b[39;00m\n\u001b[1;32m      3\u001b[0m w \u001b[38;5;241m=\u001b[39m (gc[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchr\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mch\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m&\u001b[39m (gc[\u001b[38;5;241m3\u001b[39m] \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m pos) \u001b[38;5;241m&\u001b[39m (gc[\u001b[38;5;241m4\u001b[39m] \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m pos)\n\u001b[0;32m----> 5\u001b[0m TSS \u001b[38;5;241m=\u001b[39m pos \u001b[38;5;241m-\u001b[39m \u001b[43mgc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mw\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      6\u001b[0m TES \u001b[38;5;241m=\u001b[39m gc\u001b[38;5;241m.\u001b[39mloc[w,\u001b[38;5;241m4\u001b[39m]\u001b[38;5;241m.\u001b[39mvalues[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m-\u001b[39m pos\n",
      "\u001b[0;31mIndexError\u001b[0m: index 0 is out of bounds for axis 0 with size 0"
     ]
    }
   ],
   "source": [
    "# gc = pd.read_table(gencode, header=None)\n",
    "# gc = gc.set_index([0,3,4],drop=False)\n",
    "w = (gc[0] == f\"chr{ch}\") & (gc[3] <= pos) & (gc[4] >= pos)\n",
    "\n",
    "TSS = pos - gc.loc[w,3].values[0]\n",
    "TES = gc.loc[w,4].values[0] - pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "76df060e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">chr1</th>\n",
       "      <th>65419</th>\n",
       "      <th>65433</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65520</th>\n",
       "      <th>65573</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69037</th>\n",
       "      <th>71585</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450740</th>\n",
       "      <th>451678</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284733.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685716</th>\n",
       "      <th>686654</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248913816</th>\n",
       "      <th>248913879</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248913816</td>\n",
       "      <td>248913879</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185220.12\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248916602</th>\n",
       "      <th>248919146</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248916602</td>\n",
       "      <td>248919146</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185220.12\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248906372</th>\n",
       "      <th>248906466</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248906372</td>\n",
       "      <td>248906466</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185220.12\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248913816</th>\n",
       "      <th>248913879</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248913816</td>\n",
       "      <td>248913879</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185220.12\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248917279</th>\n",
       "      <th>248917401</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248917279</td>\n",
       "      <td>248917401</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185220.12\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132133 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0       1     2          3          4  5  6  7  \\\n",
       "0    3         4                                                              \n",
       "chr1 65419     65433      chr1  HAVANA  exon      65419      65433  .  +  .   \n",
       "     65520     65573      chr1  HAVANA  exon      65520      65573  .  +  .   \n",
       "     69037     71585      chr1  HAVANA  exon      69037      71585  .  +  .   \n",
       "     450740    451678     chr1  HAVANA  exon     450740     451678  .  -  .   \n",
       "     685716    686654     chr1  HAVANA  exon     685716     686654  .  -  .   \n",
       "...                        ...     ...   ...        ...        ... .. .. ..   \n",
       "     248913816 248913879  chr1  HAVANA  exon  248913816  248913879  .  +  .   \n",
       "     248916602 248919146  chr1  HAVANA  exon  248916602  248919146  .  +  .   \n",
       "     248906372 248906466  chr1  HAVANA  exon  248906372  248906466  .  +  .   \n",
       "     248913816 248913879  chr1  HAVANA  exon  248913816  248913879  .  +  .   \n",
       "     248917279 248917401  chr1  HAVANA  exon  248917279  248917401  .  +  .   \n",
       "\n",
       "                                                                          8  \n",
       "0    3         4                                                             \n",
       "chr1 65419     65433      gene_id \"ENSG00000186092.7\"; transcript_id \"EN...  \n",
       "     65520     65573      gene_id \"ENSG00000186092.7\"; transcript_id \"EN...  \n",
       "     69037     71585      gene_id \"ENSG00000186092.7\"; transcript_id \"EN...  \n",
       "     450740    451678     gene_id \"ENSG00000284733.2\"; transcript_id \"EN...  \n",
       "     685716    686654     gene_id \"ENSG00000284662.2\"; transcript_id \"EN...  \n",
       "...                                                                     ...  \n",
       "     248913816 248913879  gene_id \"ENSG00000185220.12\"; transcript_id \"E...  \n",
       "     248916602 248919146  gene_id \"ENSG00000185220.12\"; transcript_id \"E...  \n",
       "     248906372 248906466  gene_id \"ENSG00000185220.12\"; transcript_id \"E...  \n",
       "     248913816 248913879  gene_id \"ENSG00000185220.12\"; transcript_id \"E...  \n",
       "     248917279 248917401  gene_id \"ENSG00000185220.12\"; transcript_id \"E...  \n",
       "\n",
       "[132133 rows x 9 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.loc[gc[0] == \"chr1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45fbcc10",
   "metadata": {},
   "source": [
    "### phyloP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22bde2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9gb file\n",
    "!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP100way/hg38.phyloP100way.bw\n",
    "!wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bigWigToBedGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e6620c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod +x bigWigToBedGraph\n",
    "!./bigWigToBedGraph hg38.phyloP100way.bw hg38.phyloP100way.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a83f3b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import argparse\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import pyBigWig as pbw\n",
    "\n",
    "def filelen(file):\n",
    "    fname = file.split(\"/\")[-1]\n",
    "    os.system(f\"wc -l {file} > /tmp/wc_{fname}\")\n",
    "    n = int(open(f\"/tmp/wc_{fname}\", \"r\").readlines()[0].split()[0])\n",
    "    return n\n",
    "\n",
    "def add_phyloP(df, phylop):\n",
    "    df = df.sort_values(['Chromosome', 'Position'])\n",
    "    df = df.set_index([\"Chromosome\", \"Position\"], drop=False)\n",
    "    bw = pbw.open(phylop)\n",
    "    \n",
    "    def _query(row): \n",
    "        ch, pos = row.name\n",
    "        return bw.values(f\"chr{ch}\", pos, pos+1)[0]\n",
    "\n",
    "    pp = df.apply(_query, \n",
    "             axis=1, \n",
    "             result_type='expand')\n",
    "    df.insert(4, \"pp\", pp)\n",
    "    return df\n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    # argParser = argparse.ArgumentParser()\n",
    "    # argParser.add_argument(\"--pop\", default=\"ESN\", type=str)\n",
    "    # argParser.add_argument(\"--data_dir\", default='/oak/stanford/groups/smontgom/erobb/data/watershed', type=str)\n",
    "    # argParser.add_argument(\"--gencode\", default=\"/oak/stanford/groups/smontgom/erobb/data/watershed/gencode.v43.chr_patch_hapl_scaff.annotation.exons.protein_lincRNA.gtf\", type=str)\n",
    "    # args = argParser.parse_args()\n",
    "    # pop = args.pop\n",
    "\n",
    "    pop = \"ESN\"\n",
    "    phylop = '/oak/stanford/groups/smontgom/erobb/data/hg38.phyloP100way.bw'\n",
    "    data_dir = '/oak/stanford/groups/smontgom/erobb/data/watershed'\n",
    "    tsv_in = f'AF.all.{pop}.hg38a.ID.ba.VEP.rare.ws.tsv'\n",
    "    tsv_out =  f'AF.all.{pop}.hg38a.ID.ba.VEP.gencode.phyloP.rare.ws.tsv'\n",
    "    tsv_file = f'{data_dir}/{tsv_in}'\n",
    "    tsv_file_out = f'{data_dir}/{tsv_out}'\n",
    "\n",
    "    var_df = pd.read_table(tsv_file)\n",
    "    \n",
    "    pp_df = add_phyloP(var_df, phylop)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3f725e3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SubjectID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>pp</th>\n",
       "      <th>AF</th>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <th>TF_binding_site_variant</th>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <th>...</th>\n",
       "      <th>priPhCons</th>\n",
       "      <th>mamPhCons</th>\n",
       "      <th>verPhCons</th>\n",
       "      <th>priPhyloP</th>\n",
       "      <th>mamPhyloP</th>\n",
       "      <th>verPhyloP</th>\n",
       "      <th>GerpN</th>\n",
       "      <th>GerpS</th>\n",
       "      <th>PHRED</th>\n",
       "      <th>eOutlier</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1</th>\n",
       "      <th>597782</th>\n",
       "      <td>HG03265</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>1</td>\n",
       "      <td>597782</td>\n",
       "      <td>0.953</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.297</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5.241</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597782</th>\n",
       "      <td>HG03267</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>1</td>\n",
       "      <td>597782</td>\n",
       "      <td>0.953</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.297</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5.241</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597782</th>\n",
       "      <td>HG02976</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>1</td>\n",
       "      <td>597782</td>\n",
       "      <td>0.953</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.297</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5.241</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630096</th>\n",
       "      <td>HG03301</td>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>1</td>\n",
       "      <td>630096</td>\n",
       "      <td>-1.431</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.492</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630096</th>\n",
       "      <td>HG03301</td>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>1</td>\n",
       "      <td>630096</td>\n",
       "      <td>-1.431</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.492</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">5</th>\n",
       "      <th>60891442</th>\n",
       "      <td>HG02973</td>\n",
       "      <td>ENSG00000049167</td>\n",
       "      <td>5</td>\n",
       "      <td>60891442</td>\n",
       "      <td>-0.127</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.682</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.321</td>\n",
       "      <td>0.504</td>\n",
       "      <td>0.487</td>\n",
       "      <td>4.210</td>\n",
       "      <td>1.800</td>\n",
       "      <td>9.267</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60893612</th>\n",
       "      <td>HG03372</td>\n",
       "      <td>ENSG00000049167</td>\n",
       "      <td>5</td>\n",
       "      <td>60893612</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.471</td>\n",
       "      <td>0.477</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.642</td>\n",
       "      <td>0.642</td>\n",
       "      <td>7.065</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60893612</th>\n",
       "      <td>HG03297</td>\n",
       "      <td>ENSG00000049167</td>\n",
       "      <td>5</td>\n",
       "      <td>60893612</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.471</td>\n",
       "      <td>0.477</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.642</td>\n",
       "      <td>0.642</td>\n",
       "      <td>7.065</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60893612</th>\n",
       "      <td>HG03117</td>\n",
       "      <td>ENSG00000049167</td>\n",
       "      <td>5</td>\n",
       "      <td>60893612</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.471</td>\n",
       "      <td>0.477</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.642</td>\n",
       "      <td>0.642</td>\n",
       "      <td>7.065</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60893612</th>\n",
       "      <td>HG03366</td>\n",
       "      <td>ENSG00000049167</td>\n",
       "      <td>5</td>\n",
       "      <td>60893612</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.471</td>\n",
       "      <td>0.477</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.642</td>\n",
       "      <td>0.642</td>\n",
       "      <td>7.065</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650090 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SubjectID         GeneName  Chromosome  Position     pp  \\\n",
       "Chromosome Position                                                           \n",
       "1          597782     HG03265  ENSG00000230021           1    597782  0.953   \n",
       "           597782     HG03267  ENSG00000230021           1    597782  0.953   \n",
       "           597782     HG02976  ENSG00000230021           1    597782  0.953   \n",
       "           630096     HG03301  ENSG00000237973           1    630096 -1.431   \n",
       "           630096     HG03301  ENSG00000198744           1    630096 -1.431   \n",
       "...                       ...              ...         ...       ...    ...   \n",
       "5          60891442   HG02973  ENSG00000049167           5  60891442 -0.127   \n",
       "           60893612   HG03372  ENSG00000049167           5  60893612  0.520   \n",
       "           60893612   HG03297  ENSG00000049167           5  60893612  0.520   \n",
       "           60893612   HG03117  ENSG00000049167           5  60893612  0.520   \n",
       "           60893612   HG03366  ENSG00000049167           5  60893612  0.520   \n",
       "\n",
       "                       AF  3_prime_UTR_variant  5_prime_UTR_variant  \\\n",
       "Chromosome Position                                                   \n",
       "1          597782    0.01                    0                    0   \n",
       "           597782    0.01                    0                    0   \n",
       "           597782    0.01                    0                    0   \n",
       "           630096    0.01                    0                    0   \n",
       "           630096    0.01                    0                    0   \n",
       "...                   ...                  ...                  ...   \n",
       "5          60891442  0.01                    0                    0   \n",
       "           60893612  0.01                    0                    0   \n",
       "           60893612  0.01                    0                    0   \n",
       "           60893612  0.01                    0                    0   \n",
       "           60893612  0.01                    0                    0   \n",
       "\n",
       "                     TF_binding_site_variant  downstream_gene_variant  ...  \\\n",
       "Chromosome Position                                                    ...   \n",
       "1          597782                          0                        1  ...   \n",
       "           597782                          0                        1  ...   \n",
       "           597782                          0                        1  ...   \n",
       "           630096                          0                        0  ...   \n",
       "           630096                          0                        0  ...   \n",
       "...                                      ...                      ...  ...   \n",
       "5          60891442                        0                        1  ...   \n",
       "           60893612                        0                        1  ...   \n",
       "           60893612                        0                        1  ...   \n",
       "           60893612                        0                        1  ...   \n",
       "           60893612                        0                        1  ...   \n",
       "\n",
       "                     priPhCons  mamPhCons  verPhCons  priPhyloP  mamPhyloP  \\\n",
       "Chromosome Position                                                          \n",
       "1          597782        0.046      0.046      0.297      0.000      0.000   \n",
       "           597782        0.046      0.046      0.297      0.000      0.000   \n",
       "           597782        0.046      0.046      0.297      0.000      0.000   \n",
       "           630096        0.007      0.007      0.007      0.000      0.000   \n",
       "           630096        0.007      0.007      0.007      0.000      0.000   \n",
       "...                        ...        ...        ...        ...        ...   \n",
       "5          60891442      0.682      0.001      0.001      0.321      0.504   \n",
       "           60893612      0.162      0.471      0.477      0.404      0.698   \n",
       "           60893612      0.162      0.471      0.477      0.404      0.698   \n",
       "           60893612      0.162      0.471      0.477      0.404      0.698   \n",
       "           60893612      0.162      0.471      0.477      0.404      0.698   \n",
       "\n",
       "                     verPhyloP  GerpN  GerpS  PHRED  eOutlier  \n",
       "Chromosome Position                                            \n",
       "1          597782        0.000  0.000  0.000  5.241       0.0  \n",
       "           597782        0.000  0.000  0.000  5.241       0.0  \n",
       "           597782        0.000  0.000  0.000  5.241       0.0  \n",
       "           630096        0.000  0.000  0.000  3.492       0.0  \n",
       "           630096        0.000  0.000  0.000  3.492       0.0  \n",
       "...                        ...    ...    ...    ...       ...  \n",
       "5          60891442      0.487  4.210  1.800  9.267       0.0  \n",
       "           60893612      0.675  0.642  0.642  7.065       0.0  \n",
       "           60893612      0.675  0.642  0.642  7.065       0.0  \n",
       "           60893612      0.675  0.642  0.642  7.065       0.0  \n",
       "           60893612      0.675  0.642  0.642  7.065       0.0  \n",
       "\n",
       "[650090 rows x 43 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "752cfbf5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
